### PR TITLE
[ADD] [hw_]l10n_ke_edi_tremol: Modules for communicating with tremol fiscal device

### PR DESCRIPTION
**Pull request by dbkosky**

### CODE REVIEW

**Comment by pimodoo** [comment]
@antonylesuisse @jco-odoo @dbkosky @qle-odoo Hello, I think this module should be put in enterprise, as it will use IOT devices in near future.

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+
+class TremolG03Driver(SerialDriver):
+    """Driver for the blackbox fiscal data module."""
+
+    _protocol = TremolG03Protocol
+
+    def __init__(self, identifier, device):
+        super().__init__(identifier, device)
+        self.device_type = 'ke_fiscal_device'
```
Comment (qle-odoo) [inline addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:54]: `self.device_type = 'fiscal_data_module'`

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        :param device: path to the device
+        :type device: str
+        :return: whether the device is supported by the driver
+        :rtype: bool
+        """
+        protocol = cls._protocol
+        try:
+            protocol = cls._protocol
+            with serial_connection(device['identifier'], protocol) as connection:
+                connection.write(b'\x04')
```
Comment (qle-odoo) [inline addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:75]: You can use `connection.write(b'\x09')`  and `response == b'\x40'` it is more specific to your device
check `1.3. SHORT MESSAGES FOR TESTING THE STATUS OF THE CU`

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            core_message = struct.pack('BB%ds' % (len(msg)), len(msg)+34, self.message_number+32, msg)
+            request = struct.pack('B%ds2sB' % (len(core_message)), 2, core_message, self.generate_checksum(core_message), 10)
+            with self._device_lock:
+                time.sleep(self._protocol.commandDelay)
+                self._connection.write(request)
+                time.sleep(self._protocol.measureDelay)
+                response = self._connection.read_all()
+            if response[0] == 0x06:
+                if response[2] != 0x30 and response[3] != 0x30:
+                    raise Exception('Error %s %s', response[2], response[3])
```
Comment (qle-odoo) [inline addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:138]: Maybe you can return the human readable error ?
convert this array to a json
`1.2.1. ACKNOWLEDGEMENT RESPONSE`
Like: https://github.com/odoo/enterprise/blob/16.0/pos_l10n_se/iot_handlers/drivers/SerialPosBlackBoxSweden.py#L43

**Review by jco-odoo — COMMENTED** [review]
I need to check further still.  For the sysparam, it might be not practical this way to have multi-company with each their device anyways, but that is not why you do it.

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            protocol = cls._protocol
+            with serial_connection(device['identifier'], protocol) as connection:
+                connection.write(b'\x09')
+                time.sleep(protocol.commandDelay)
+                response = connection.read(1)
+                if response == b'\x40':
+                    return True
+
+        except serial.serialutil.SerialTimeoutException:
+            pass
```
[the remark this answers is not in the document]
Comment (dbkosky) [reply addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:101]: `pass` is used elsewhere, in the SerialScaleDrivers ..

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+        :returns: a list containing one byte-string representing the <CMD> and
+                  <DATA> of the message sent to the fiscal device.
+        """
+        headquarter_address = (self.commercial_partner_id.street or '') + (self.commercial_partner_id.street2 or '')
+        customer_address = (self.partner_id.street or '') + (self.partner_id.street2 or '')
+        postcode_and_city = self.partner_id.zip or '' + ' ' +  self.partner_id.city or ''
+        invoice_elements = [
+            b'1',                                                   # Reserved - 1 symbol with value '1'
+            b'     0',                                              # Reserved - 1 symbol with value ‘     0’
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:87]: rjust?

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            b'0',                                                   # Reserved - 1 symbol with value '0'
+            b'1' if self.move_type == 'out_invoice' else b'A',      # 1 symbol with value '1' (new invoice), 'A' (credit note), or '@' (debit note)
+            self._l10n_ke_fmt(self.commercial_partner_id.name, 30), # 30 symbols for Company name
+            self._l10n_ke_fmt(self.commercial_partner_id.vat, 14),  # 14 Symbols for the client PIN number
+            self._l10n_ke_fmt(headquarter_address, 30),             # 30 Symbols for customer headquarters
+            self._l10n_ke_fmt(customer_address, 30),                # 30 Symbols for the address
+            self._l10n_ke_fmt(postcode_and_city, 30),               # 30 symbols for the customer post code and city
+            self._l10n_ke_fmt('', 30),                              # 30 symbols for the exemption number
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:95]: why did you remove exemption number?  Did not we have it on the partner?  People should be able to report it...

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        ]
+        if self.move_type in ('out_refund', 'debit_note'):
+            invoice_elements.append(self._l10n_ke_fmt(self.reversed_entry_id.l10n_ke_cu_invoice_number, 19)), # 19 symbols for related invoice number
+        invoice_elements.append(self._l10n_ke_fmt(self.name, 15))   # 15 symbols for trader system invoice number
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:99]: Is there a way to have [15:] instead of [:15] for this?  (the number info is mostly in the end...)

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+            return set(discount_taxes.ids) == set(other_line_taxes.ids)
+
+        lines = self.invoice_line_ids.filtered(lambda l: not l.display_type and l.quantity and l.price_total)
+        discount_dict = {line.id: line.discount for line in lines if line.price_total > 0}
+        for line in lines:
+            if not is_discount_line(line):
+                continue
+            # Search for non-discount lines
+            candidate_vals_list = [l for l in lines if not is_discount_line(l) and is_candidate(l, line)]
+            candidate_vals_list = sorted(candidate_vals_list, key=lambda x: x.price_unit * x.quantity, reverse=True)
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:133]: why not price_subtotal?

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+                self._l10n_ke_fmt(hscode, 10),                  # 10 symbols for HS code in the format xxxx.xx.xx (can be empty)
+                self._l10n_ke_fmt(hsname, 20),                  # 20 symbols for the HS name (can be empty)
+                str(percentage).encode('cp1251')[:5]             # up to 5 symbols for vat rate
+            ])
+
+            # 1 to 10 symbols for quantity
+            line_data += b'*' + str(line.quantity).encode('cp1251')[:10]
+            if discount_dict[line.id]:
+                # 1 to 7 symbols for percentage of discount/addition
+                discount_sign = b'-' if discount_dict[line.id] > 0 else b'+'
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:181]: You confirm that if discount 0.0, we do not need to add anything?

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                report_line_ids = line.tax_ids[0].invoice_repartition_line_ids.filtered(lambda r: r.repartition_type == 'base').tag_ids.tax_report_line_ids.ids
+                letter = 'E' if self.env.ref('l10n_ke.tax_report_line_exempt_sales_base', raise_if_not_found=False).id in report_line_ids else 'C'
+
+            uom = line.product_uom_id and line.product_uom_id.name or ''
+            if letter != 'A':
+                hscode = line.product_id.l10n_ke_hsn_code
+                hsname = line.product_id.l10n_ke_hsn_name
+            else:
+                hscode = ''
+                hsname = ''
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:166]:
```suggestion
              hscode = line.product_id.l10n_ke_hsn_code if letter !='A' else ''
              hsname = line.product_id.l10n_ke_hsn_name letter !='A' else ''
```

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        # CMD_CLOSE_FISCAL_RECEIPT = 0x38
+        msgs += [b'\x38']
+        return msgs
+
+    # -------------------------------------------------------------------------
+    # POST COMMANDS / RECEIVE DATA
+    # -------------------------------------------------------------------------
+
+    def l10n_ke_action_cu_post(self):
+        self.ensure_one()
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:207]: Why not put the validation here?

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+                # 1 to 7 symbols for percentage of discount/addition
+                discount_sign = b'-' if discount_dict[line.id] > 0 else b'+'
+                discount = discount_sign + str(abs(discount_dict[line.id])).encode('cp1251')[:6]
+                line_data += b',' + discount + b'%'
+
+            # Command: Sale of article (0x31)
+            msgs += [b'\x31' + line_data]
+        return msgs
+
+    def _l10n_ke_cu_message(self):
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:189]:
```suggestion
    def _l10n_ke_get_cu_messages(self):
```

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+        :param messages:     The messages (consisting of <CMD> and <DATA>) to
+                             send to the fiscal device.
+        :returns:            Dictionary containing a list of the responses from
+                             fiscal device and status of the fiscal device.
+        """
+        device = TremolG03Driver.get_default_device()
+        if device:
+            # First run the command to get the fiscal device numbers
+            device_numbers = device.send([b'\x60'])
```
Comment (jco-odoo) [inline addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:218]: I guess we do not add this to the regular messages as it does not require abort?

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        :return:     A list of the responses (if any) from the device. If the
+                     response is an ack, it wont be part of this list.
+        """
+
+        replies = []
+        for msg in msgs:
+            self.message_number += 1
+            core_message = struct.pack('BB%ds' % (len(msg)), len(msg)+34, self.message_number+32, msg)
+            request = struct.pack('B%ds2sB' % (len(core_message)), 2, core_message, self.generate_checksum(core_message), 10)
+            with self._device_lock:
```
[the remark this answers is not in the document]
Comment (dbkosky) [reply addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:149]: to test

File: addons/l10n_ke/data/account_tax_report_data.xml [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+            <record id="tax_report_line_sales_tax" model="account.tax.report.line">
+                <field name="name">Details of Sales (Tax)</field>
+                <field name="sequence">2</field>
+                <field name="formula">None</field>
+                <field name="children_line_ids">
+                    <record id="tax_report_line_general_rate_sales_tax" model="account.tax.report.line">
+                        <field name="name">1. General Rate 16% (Tax)</field>
+                        <field name="tag_name">1 General Rate Tax</field>
+                        <field name="code">box_1_tax</field>
+                        <field name="sequence">1</field>
```
[the remark this answers is not in the document]
Comment (dbkosky) [reply addons/l10n_ke/data/account_tax_report_data.xml:57]: that's in 16

File: addons/l10n_ke_edi_tremol/models/account_move.py (hunk 8 above)
Comment (dbkosky) [reply addons/l10n_ke_edi_tremol/models/account_move.py:133]: not sure what you're referring to here

File: addons/l10n_ke_edi_tremol/models/account_move.py (hunk 9 above)
Comment (dbkosky) [reply addons/l10n_ke_edi_tremol/models/account_move.py:181]: image

yes it's optional

File: addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+                replies.append('')
+            elif response[0] == 0x15:
+                self.data['status'] = "Received NACK"
+                _logger.error("Sent request: %s,\n Received NACK \x15", msg)
+                self.abort_post()
+                break
+            elif response[0] == 0x02:
+                self.data['status'] = "ok"
+                size = response[1] - 35
+                reply = response[4:4+size]
```
Comment (jco-odoo) [inline addons/hw_drivers/iot_handlers/drivers/L10nKeEDISerialDriver.py:179]:
```suggestion
                reply = response[4:4 + size]
```

File: addons/l10n_ke_edi_tremol/__manifest__.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        'views/account_move_view.xml',
+        'views/product_view.xml',
+        'views/report_invoice.xml',
+        'views/res_config_settings_view.xml',
+        'views/res_partner_views.xml',
+    ],
+    'assets': {
+        'web.assets_backend': [
+            'l10n_ke_edi_tremol/static/src/js/send_invoice.js',
+            'l10n_ke_edi_tremol/static/src/js/get_device_info.js',
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/__manifest__.py:25]: @dbkosky This line does not seem to be necessary anymore

File: addons/l10n_ke_edi_tremol/__manifest__.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': "Kenya Tremol Device EDI Integration",
+    'summary': """
+            Kenya Tremol Device EDI Integration
+        """,
+    'description': """
+       This module integrates with the Kenyan G03 tremol control unit device.
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/__manifest__.py:8]:
```suggestion
       This module integrates with the Kenyan G03 Tremol control unit device for sending invoices to TIMS.
```

File: addons/l10n_ke_edi_tremol/models/account_move.py [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+            # Letter to classify tax, 0% taxes are handled conditionally, as the tax can be zero-rated or exempt
+            letter = ''
+            if percentage in vat_class:
+                letter = vat_class[percentage]
+            else:
+                report_line_ids = line.tax_ids[0].invoice_repartition_line_ids.filtered(lambda r: r.repartition_type == 'base').tag_ids.tax_report_line_ids.ids
+                try:
+                    exempt_report_line = self.env.ref('l10n_ke.tax_report_line_exempt_sales_base', raise_if_not_found=False).id
+                except ValueError:
+                    raise UserError(_("Tax a exempt report line cannot be found, please update the l10n_ke module."))
```
Comment (jco-odoo) [inline addons/l10n_ke_edi_tremol/models/account_move.py:155]:
```suggestion
                    raise UserError(_("Tax exempt report line cannot be found, please update the l10n_ke module."))
```

**Comment by dbkosky** [comment]
successfully merged in 
https://github.com/odoo/odoo/pull/106653
